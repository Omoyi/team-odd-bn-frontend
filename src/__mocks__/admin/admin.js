import {
  ASSIGN_ROLE_SUCCESS,
  GET_ALL_ROLES,
  ADMIN_ERROR,
  NEW_ROLE_INPUT,
} from '../../redux/actionTypes/adminActionTypes';

export const retrievedRolesApi = {
  status: 200,
  response: {
    message: 'All roles retrieved successfully',
    data: [
      {
        id: 1,
        type: 'super admin',
      },
    ],
  },
};

export const retrievedRolesAction = [
  {
    type: GET_ALL_ROLES,
    allRoles: {
      message: 'All roles retrieved successfully',
      data: [
        {
          id: 1,
          type: 'super admin',
        },
      ],
    },
  },
  {
    type: ADMIN_ERROR,
    status: 'success',
    error: {},
  },
];

export const assignRolesApi = {
  status: 201,
  response: {
    message: 'Role assigned successfully',
    data: {
      role: 'super admin',
    },
  },
};

export const assignRolesAction = [
  {
    type: ASSIGN_ROLE_SUCCESS,
    payload: {
      message: 'Role assigned successfully',
      data: {
        role: 'super admin',
      },
    },
  },
  {
    type: ADMIN_ERROR,
    status: 'success',
    error: {},
  },
];

export const assignRolesError = {
  status: 404,
  response: {
    data: {
      message: 'Email does not exist',
    },
  },
};

export const assignRolesErrorAction = [
  {
    type: ADMIN_ERROR,
    status: 'error',
    error: {
      message: 'Email does not exist',
    },
  },
];

export const rolenInputAction = {
  type: NEW_ROLE_INPUT,
  newRoleId: 5,
};

export const adminError = {
  type: ADMIN_ERROR,
  error: 'Role is already assigned',
};
export const getAllRoles = {
  type: GET_ALL_ROLES,
  allRoles: [
    {
      id: 1,
      type: 'super admin',
    },
  ],
};
export const assignRole = {
  type: ASSIGN_ROLE_SUCCESS,
  payload: {
    data: {
      message: 'Role assigned successfully',
      data: {
        role: 'super admin',
      },
    },
  },
};

export const defaultProps = {
  adminState: {
    admin: {},
    allUsers: {
      users: {
        message: '',
        data: [
          {
            id: 1,
            email: 'me@you.com',
            firstName: 'me',
            lastName: 'you',
            roleId: 3,
            role: {
              type: 'user',
            },
          },
        ],
      },
    },
    search: {
      payload: [
        {
          id: 1,
          email: 'me@you.com',
          firstName: 'me',
          lastName: 'you',
          roleId: 3,
          role: {
            type: 'user',
          },
        },
      ],
    },
  },
  itemsPerPage: 1,
  users: [
    {
      id: 50,
      firstName: 'Victor',
      lastName: 'KARANGWA',
      email: 'victor.karangwa@andela.com',
      roleId: 3,
      role: {
        type: 'user',
      },
    },
    {
      id: 16,
      firstName: 'Wlly 2',
      lastName: 'Boris 2',
      email: 'sugiraw@gmail.com',
      roleId: 3,
      role: {
        type: 'user',
      },
    },
  ],
  getAllUsers: jest.fn(),
  getRolesAction: jest.fn(),
  assignRole: jest.fn(),
  searchResults: jest.fn(),
};

export const updateDefaultProps = {
  adminState: {
    admin: {
      payload: {
        message: 'All roles retrieved successfully',
      },
      allRoles: {
        data: [
          {
            id: 3,
            type: 'user',
          },
        ],
      },
    },
    allUsers: {
      users: {
        message: '',
        data: [
          {
            id: 1,
            email: 'me@you.com',
            firstName: 'me',
            lastName: 'you',
            roleId: 3,
            role: {
              type: 'user',
            },
          },
        ],
      },
    },
  },
  getAllUsers: jest.fn(),
  getRolesAction: jest.fn(),
  assignRole: jest.fn(),
};

export const emptyDataComponent = {
  adminState: {
    admin: {
      payload: 'data',
      allRoles: [1, 2, 3],
    },
    allUsers: {
      users: {
        message: '',
        data: '',
      },
    },
  },
  users: [
    {
      id: 50,
      firstName: 'Victor',
      lastName: 'KARANGWA',
      email: 'victor.karangwa@andela.com',
      roleId: 3,
      role: {
        type: 'user',
      },
    },
    {
      id: 16,
      firstName: 'Wlly 2',
      lastName: 'Boris 2',
      email: 'sugiraw@gmail.com',
      roleId: 3,
      role: {
        type: 'user',
      },
    },

  ],
  getAllUsers: jest.fn(),
  getRolesAction: jest.fn(),
  assignRole: jest.fn(),
};

export const options = [
  { id: 6, type: 'manager' },
  { id: 3, type: 'user' },
];
