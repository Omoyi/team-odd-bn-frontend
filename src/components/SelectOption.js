import React from 'react';
import '../assets/css/selectOption.scss';

const SelectOption = ({
  classname, options, handleChange, email,
}) => {
  return (
    <select data-test="selectOption-test" defaultValue="" className={classname} onChange={(e) => (handleChange(e.target.value, email))}>
      <option disabled value="">
      Choose
      </option>
      {options
      && options.map((item) => (
        <option value={item.id} key={item.id}>
          {item.type}
        </option>
      ))}
    </select>
  );
};

export default SelectOption;
