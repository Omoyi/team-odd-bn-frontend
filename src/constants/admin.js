import React from 'react';
import { Link, BrowserRouter as Router } from 'react-router-dom';
import { connect } from 'react-redux';
import { handleFilter } from '../helpers/filter';
import { checkPrevilege } from '../helpers/checkPrevilege';
import usersIcon from '../assets/images/users.png';

export const tableHeads = [
  { className: 'fNameNav', attribute: 'First name', key: 'firstName' },
  { className: 'lNameNav', attribute: 'Last name', key: 'lastName' },
  { className: 'emailNav', attribute: 'Email', key: 'email' },
  { className: 'role', attribute: 'Role', key: 'role' },
  { className: 'EditRoleNav', attribute: 'Edit role', key: 'editRole' },
];
export const userRoleNav = (data) => {
  let newData;
  const roleNavs = [
    {
      className: 'tableHeader',
      attribute: <h2>Assign Roles</h2>,
      key: 'tableHeader',
    },
  ];
  return { roleNavs, newData };
};

export const usersNav = () => checkPrevilege(1) && (
<div>
  <img src={usersIcon} alt="Users icon" />
  <span>Users</span>
</div>
);
export const mapStateToProps = (state) => ({
  admin: state.allUsers.users,
  pageNo: state.pagination.pageNo,
  itemsPerPage: state.pagination.itemsPerPage,
});

export default connect(mapStateToProps, {})(userRoleNav);
