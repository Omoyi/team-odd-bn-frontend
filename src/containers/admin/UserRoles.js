import React, { Component } from 'react';
import { connect } from 'react-redux';
import AllUsers from './allUsers';
import {
  paginationAction, getRolesAction, newRoleInput, assignRole,
} from '../../redux/actions/admin/adminAction';

import getAllUsers from '../../redux/actions/getAllUsersActions';

export class UserRoles extends Component {
  async componentDidMount() {
    await this.props.getAllUsers();
    this.props.getRolesAction();
  }

  componentDidUpdate() {
    this.props.adminState.admin.payload && this.props.getAllUsers();
    this.props.adminState.admin.payload = undefined;
  }

  render() {
    const data = this.props.adminState.allUsers.users
      && this.props.adminState.allUsers.users.data;
    const options = this.props.adminState.admin.allRoles
      && this.props.adminState.admin.allRoles.data;

    return (
      <AllUsers
        paginatedRequest={this.props.adminState.allUsers.paginatedRequest}
        paginationAction={this.props.paginationAction}
        paginationEnd={this.props.adminState.allUsers.paginationEnd}
        paginationStart={this.props.adminState.allUsers.paginationStart}
        data-test="userRoles-test"
        handleChange={this.props.assignRole}
        data={data}
        options={options}
      />
    );
  }
}
export const mapStateToProps = (state) => ({
  adminState: state,
});
export default connect(mapStateToProps, {
  getAllUsers,
  getRolesAction,
  newRoleInput,
  assignRole,
  paginationAction,

})(UserRoles);
