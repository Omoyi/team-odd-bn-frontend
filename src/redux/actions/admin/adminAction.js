import {
  ASSIGN_ROLE_SUCCESS, GET_ALL_ROLES, ADMIN_ERROR, NEW_ROLE_INPUT,
} from '../../actionTypes/adminActionTypes';
import { PAGINATION } from '../../actionTypes/profileActionTypes';
import {
  notificationError,
  notificationSuccess,
} from '../../../helpers/notificationPopUp';
import apiCall from '../../../helpers/apiCall';

const emptyError = {
  type: ADMIN_ERROR,
  status: 'success',
  error: {},
};
const triggerError = (error) => ({
  type: ADMIN_ERROR,
  status: 'error',
  error: error.response.data,
});
export const getRolesAction = () => async (dispatch) => {
  const HEADERS_REQUEST = {
    token: localStorage.getItem('token'),
  };
  const roles = await apiCall.get('/users/role', {
    headers: HEADERS_REQUEST,
  });
  dispatch({
    type: GET_ALL_ROLES,
    allRoles: roles.data,
  });
  dispatch(emptyError);
};
export const assignRole = (roleId, email) => async (dispatch) => {
  const HEADERS_REQUEST = {
    token: localStorage.getItem('token'),
  };
  try {
    const assignNewRole = await apiCall.put(`/users/role/${roleId}`, { email }, {
      headers: HEADERS_REQUEST,
    });
    dispatch({
      type: ASSIGN_ROLE_SUCCESS,
      payload: assignNewRole.data,
    });
    notificationSuccess(assignNewRole.data.message);
    dispatch(emptyError);
  } catch (error) {
    dispatch(triggerError(error));
    notificationError(error.response.data.message);
  }
};

export const newRoleInput = (newRoleId) => ({
  type: NEW_ROLE_INPUT,
  newRoleId,
});


export const paginationAction = (data) => {
  if (data.data.length < data.paginationEnd) {
    data.paginationStart = 0;
    data.paginationEnd = 5;
  }
  if (data.paginationEnd < 5) {
    data.paginationStart = data.data.length - 5;
    data.paginationEnd = data.data.length;
  }
  return {
    type: PAGINATION,
    payload: data,
  };
};
