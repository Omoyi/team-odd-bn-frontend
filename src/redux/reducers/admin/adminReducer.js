import {
  ASSIGN_ROLE_SUCCESS,
  GET_ALL_ROLES,
  ADMIN_ERROR,
  NEW_ROLE_INPUT,
} from '../../actionTypes/adminActionTypes';

const initialState = {
  error: null,
};

const adminReducer = (state = initialState, action) => {
  switch (action.type) {
    case ASSIGN_ROLE_SUCCESS:
      return {
        ...state,
        payload: action.payload,
      };
    case GET_ALL_ROLES:
      return { ...state, allRoles: action.allRoles };
    case ADMIN_ERROR:
      return { ...state, error: action.error };
    case NEW_ROLE_INPUT:
      return { ...state, newRoleId: action.newRoleId };
    default:
      return state;
  }
};

export default adminReducer;
