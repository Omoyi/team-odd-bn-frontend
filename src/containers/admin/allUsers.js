import React from 'react';
import CommonTable from '../../views/trips/CommonTable';
import SelectOption from '../../components/SelectOption';
import Status from '../../components/Status';
import { userRoleNav, tableHeads } from '../../constants/admin';

const allUsers = ({
  data, options, handleChange, paginationEnd,
  paginationStart,
  paginatedRequest,
  paginationAction,
}) => {
  const entities = paginatedRequest ? paginatedRequest.map((user) => [
    { className: 'firstName', attribute: user.firstName, key: 'firstName' },
    { className: 'LastName', attribute: user.lastName, key: 'lastName' },
    { className: 'Email', attribute: user.email, key: 'email' },
    { className: 'UserRole', attribute: user.role && user.role.type, key: 'roleId' },
    { className: 'EditRole', attribute: <SelectOption handleChange={handleChange} email={user.email} options={options} classname="selectOption" />, key: 'action' },
  ]) : [];

  return (
    <CommonTable
      data={data}
      paginationAction={paginationAction}
      paginationEnd={paginationEnd}
      paginationStart={paginationStart}
      paginatedRequest={paginatedRequest}
      navs={userRoleNav}
      tableHeads={tableHeads}
      entities={entities}
    />
  );
};


export default allUsers;
