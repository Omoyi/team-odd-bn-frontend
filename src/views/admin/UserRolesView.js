import React from 'react';
import dotenv from 'dotenv';
import '../../assets/css/authentication.scss';
import UserRoles from '../../containers/admin/UserRoles';
import Dashboard from '../Dashboard/sidebar/index';

dotenv.config();

const UserRolesView = () => (
  <Dashboard data-testid="userRoles-View">
    <UserRoles />
  </Dashboard>
);
export default UserRolesView;
