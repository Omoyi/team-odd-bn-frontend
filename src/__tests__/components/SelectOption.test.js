import React from 'react';
import { shallow } from 'enzyme';
import SelectOption from '../../components/SelectOption';
import { options } from '../../__mocks__/admin/admin';

describe('Test suite for Select option component', () => {
  it('should call onChange prop for signup', () => {
    const mockFn = jest.fn();
    expect(
      shallow(
        <SelectOption
          classname="SelectOption"
          options={options}
          email="me@you.com"
          handleChange={mockFn}
        />,
      ),
    ).toMatchSnapshot();
    const select = shallow(
      <SelectOption
        classname="SelectOption"
        options={options}
        email="me@you.com"
        handleChange={mockFn}
      />,
    );
    const cc = select.find('[data-test="selectOption-test"]');
    cc.simulate('change', { target: { value: 'user' } });
    expect(mockFn.mock.calls[0][0]).toEqual('user');
  });
});
