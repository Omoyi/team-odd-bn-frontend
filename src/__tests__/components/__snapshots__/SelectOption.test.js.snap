// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test suite for Select option component should call onChange prop for signup 1`] = `ShallowWrapper {}`;
